File: test/rql_test/connections/connection.js
/////
// Tests the driver API for making connections and excercising the networking code
/////

var assert = require('assert');
var path = require('path');
var fs = require('fs');
var spawn = require('child_process').spawn

// -- load rethinkdb from the proper location

var r = require(path.resolve(__dirname, '..', 'importRethinkDB.js')).r;

// -- get ENV inforamtion

var testDefault = process.env.TEST_DEFAULT_PORT == "1";
var driverPort = process.env.RDB_DRIVER_PORT;
var serverHost = process.env.RDB_SERVER_HOST || 'localhost';

// -- helper functions

var givesError = function(type, msg, done){
    return function(err){
        var error = null;
        try {
            assertNotNull(err);
            assert.equal(err.constructor.name, type);
            var _msg = err.message.replace(/ in:\n([\r\n]|.)*/m, "");
            _msg = _msg.replace(/\nFailed assertion:(.|\n)*/m, "")
            assert.equal(_msg, msg);
        } catch (e) {
            error = e;
        } finally {
            if (done) {
                done(error);
            } else if (error) {
                throw error;
            } else {
                return true; // error is correct
            }
        }
    };
};

var withConnection = function(f){
    return function(done){
        r.connect({host:serverHost, port:driverPort}, function(err, c){
            assertNull(err);
            f(done, c);
        });
    };
};

var noError = function(f){
    return function(err){
        assertNull(err);
        f.apply({}, arguments);
    };
}

var assertNull = function(x){
    assert.equal(x, null);
}

var assertNotNull = function(x){
    assert.notEqual(x, null);
}

var ifTestDefault = function(f, c){
    if(testDefault){
        return f(c);
    }else{
        return c();
    }
}

describe('Javascript connection API', function(){
    it("times out with a server that doesn't do the handshake correctly", function(done){
        // Setup dummy sever
        var server = require('net').createServer(function(c) {});
        server.listen(0, function() {
            var badPort = server.address().port;
            r.connect({port:badPort, timeout:1}, givesError("RqlDriverError", "Handshake timedout", done));
        });
    });
    
    describe('With no server', function(){
        it("fails when trying to connect", function(done){
            ifTestDefault(
                function(cont){
                    r.connect({}, givesError("RqlDriverError", "Could not connect to localhost:28015.\nconnect ECONNREFUSED", function(){
                        r.connect({host:'0.0.0.0'}, givesError("RqlDriverError", "Could not connect to 0.0.0.0:28015.\nconnect ECONNREFUSED", cont))})); },
                function(){
                    r.connect({port:11221}, givesError("RqlDriverError", "Could not connect to localhost:11221.\nconnect ECONNREFUSED", function(){
                        r.connect({host:'0.0.0.0', port:11221}, givesError("RqlDriverError", "Could not connect to 0.0.0.0:11221.\nconnect ECONNREFUSED", done))}))});
        });
        
        it("empty run", function(done) {
            r.expr(1).run().then(function() {
                done(new Error("Was expecting an error"))
            }).error(givesError("RqlDriverError", "First argument to `run` must be an open connection.", done));
        });
        
        it("run with an object", function(done) {
            r.expr(1).run({}).then(function() {
                done(new Error("Was expecting an error"))
            }).error(givesError("RqlDriverError", "First argument to `run` must be an open connection.", done));
        });
    });
    
    describe('With a server', function(){
        this.timeout(4000)
        // TODO: test default port
        
        describe('close twice and reconnect', function(){
            var simpleQuery = function(c) { return r(1).run(c); };
            
            it("noreplyWait=false", function(done) {
                var conn = null;
                r.connect({host:serverHost, port:driverPort})
                .then(function(c) {
                    conn = c;
                    return simpleQuery(conn);
                }).then(function() {
                    return conn.close({noreplyWait: false});
                }).then(function() {
                    return conn.close({noreplyWait: false});
                }).then(function() {
                    return conn.reconnect();
                }).then(function() {
                    return simpleQuery(conn);
                }).nodeify(done)
            });
            
            it("synchronously", function(done) {
                var conn = null;
                r.connect({host:serverHost, port:driverPort})
                .then(function(c) {
                    conn = c;
                    return simpleQuery(conn);
                }).then(function() {
                    return conn.close();
                }).then(function() {
                    return conn.close();
                }).then(function() {
                    return conn.reconnect();
                }).then(function() {
                    return simpleQuery(conn);
                }).nodeify(done)
            });
        })
        
        it("fails to query after close", withConnection(function(done, c){
            c.close({noreplyWait: false});
            r(1).run(c, givesError("RqlDriverError", "Connection is closed.", done));
        }));
        
        it("noreplyWait waits", withConnection(function(done, c){
            var t = new Date().getTime();
            r.js('while(true);', {timeout: 0.5}).run(c, {noreply: true}, function() {});
            c.noreplyWait(function (err) {
                assertNull(err);
                var duration = new Date().getTime() - t;
                assert(duration >= 500);
                done();
            });
        }));
        
        it("close waits by default", withConnection(function(done, c){
            var t = new Date().getTime();
            r.js('while(true);', {timeout: 0.5}).run(c, {noreply: true}, function() {});
            c.close(function (err) {
                assertNull(err);
                var duration = new Date().getTime() - t;
                assert(duration >= 500);
                done();
            });
        }));
        
        it("reconnect waits by default", withConnection(function(done, c){
            var t = new Date().getTime();
            r.js('while(true);', {timeout: 0.5}).run(c, {noreply: true}, function() {});
            c.reconnect(function (err) {
                assertNull(err);
                var duration = new Date().getTime() - t;
                assert(duration >= 500);
                done();
            });
        }));
        
        it("close does not wait if we want it not to", withConnection(function(done, c){
            var t = new Date().getTime();
            r.js('while(true);', {timeout: 0.5}).run(c, {noreply: true}, function() {});
            c.close({'noreplyWait': false}, function (err) {
                assertNull(err);
                var duration = new Date().getTime() - t;
                assert(duration < 500);
                done();
            });
        }));
        
        it("reconnect does not wait if we want it not to", withConnection(function(done, c){
            var t = new Date().getTime();
            r.js('while(true);', {timeout: 0.5}).run(c, {noreply: true}, function() {});
            c.reconnect({'noreplyWait': false}, function (err) {
                assertNull(err);
                var duration = new Date().getTime() - t;
                assert(duration < 500);
                done();
            });
        }));
        
        it("close waits even without callback", withConnection(function(done, c){
            var start = Date.now();
            var timeout = 1.5;
            r.js('while(true);', {timeout: timeout}).run(c, {noreply: true}, function() {});
            c.close(function(err) {
                if (err) throw err;
                var duration = Date.now()-start;
                assert(duration > timeout*1000);
                done()
            });
        }));
        
        it("useOutdated", withConnection(function(done, c){
            r.db('test').tableCreate('useOutdated').run(c, function(){
                r.db('test').table('useOutdated', {useOutdated:true}).run(c, function(){
                    r.db('test').table('useOutdated').run(c, {useOutdated: true}, done);});});
        }));
        
        it("test default durability", withConnection(function(done, c){
            r.db('test').tableCreate('defaultDurability').run(c, noError(function(){
                r.db('test').table('defaultDurability').insert({data:"5"}).run(c, {durability: "default"},
                    givesError("RqlRuntimeError", "Durability option `default` unrecognized (options are \"hard\" and \"soft\")", done));
            }));
        }));
        
        it("test wrong durability", withConnection(function(done, c){
            r.db('test').tableCreate('wrongDurability').run(c, noError(function(){
                r.db('test').table('wrongDurability').insert({data:"5"}).run(c, {durability: "wrong"},
                    givesError("RqlRuntimeError", "Durability option `wrong` unrecognized (options are \"hard\" and \"soft\")", done))
            }));
        }));
        
        it("test soft durability", withConnection(function(done, c){
            r.db('test').tableCreate('softDurability').run(c, noError(function(){
                r.db('test').table('softDurability').insert({data:"5"}).run(c, {durability: "soft"}, noError(done));
            }));
        }));
        
        it("test hard durability", withConnection(function(done, c){
            r.db('test').tableCreate('hardDurability').run(c, noError(function(){
                r.db('test').table('hardDurability').insert({data:"5"}).run(c, {durability: "hard"}, noError(done));
            }));
        }));
        
        it("test non-deterministic durability", withConnection(function(done, c){
            r.db('test').tableCreate('nonDeterministicDurability').run(c, noError(function(){
                r.db('test').table('nonDeterministicDurability').insert({data:"5"}).run(c, {durability: r.js("'so' + 'ft'")}, noError(done));
            }));
        }));
        
        it("pretty-printing", function(){
            assert.equal(r.db('db1').table('tbl1').map(function(x){ return x; }).toString(),
                        'r.db("db1").table("tbl1").map(function(var_0) { return var_0; })')
        });
        
        it("connect() with default db", withConnection(function(done, c){
            r.dbCreate('foo').run(c, function(err, res){
                r.db('foo').tableCreate('bar').run(c, function(err, res){
                    c.close();
                    r.connect({host:serverHost, port:driverPort, db:'foo'}, noError(function(err, c){
                        r.table('bar').run(c, noError(done));
                    }));
                });
            });
        }));
        
        it("missing arguments cause exception", withConnection(function(done,c){
            assert.throws(function(){ c.use(); });
            done();
        }));
        
        // -- event emitter interface
        
        it("Emits connect event on connect", withConnection(function(done, c){
            c.on('connect', function(err, c2) {
                done();
            });
            
            c.reconnect(function(){});
        }));
        
        it("Emits close event on close", withConnection(function(done, c){
            c.on('close', function() {
                done();
            });
            
            c.close();
        }));
        
        // -- bad options
        
        it("Non valid optarg", withConnection(function(done, c){
            r.expr(1).run(c, {nonValidOption: true}).then(function() {
                done(new Error("Was expecting an error"))
            }).error(givesError("RqlCompileError", "Unrecognized global optional argument `non_valid_option`", done));
        }));
        
        it("Extra argument", withConnection(function(done, c){
            r.expr(1).run(c, givesError("RqlDriverError", "Second argument to `run` cannot be a function if a third argument is provided.", done), 1)
        }));
        
        it("Callback must be a function", withConnection(function(done, c){
            r.expr(1).run(c, {}, "not_a_function").error(givesError("RqlDriverError", "If provided, the callback must be a function. Please use `run(connection[, options][, callback])", done))
        }));
        
        // -- auth_key tests
        
        describe('Authorization key tests', function() {
            
            var safeConn = null;
            before(function(done) {
                r.connect({host:serverHost, port:driverPort}, function(err, conn) {
                    assertNull(err);
                    safeConn = conn;
                    done();
                });
            })
            
            afterEach(function(done) {
                // undo any auth_key changes
				r.db('rethinkdb').table('cluster_config').update({auth_key:null}).run(safeConn).then(function(value, err) { done(err); });
            })
            
            // --
            
            it("authorization key when none needed", function(done){
                r.connect(
                    {host:serverHost, port:driverPort, authKey: "hunter2"},
                    givesError("RqlDriverError", "Server dropped connection with message: \"ERROR: Incorrect authorization key.\"", done)
                );
            });
            
            it("correct authorization key", function(done){
                this.timeout(500);
                r.db('rethinkdb').table('cluster_config').update({auth_key: "hunter3"}).run(safeConn)
                .then(function() {
                    return r.connect({host:serverHost, port:driverPort, authKey: "hunter3"});
                }).then(function(authConn) {
                    return r.expr(1).run(authConn);
                }).nodeify(done)
            });
            
            it("wrong authorization key", function(done){
                this.timeout(500);
                r.db('rethinkdb').table('cluster_config').update({auth_key: "hunter4"}).run(safeConn)
                .then(function() {
                    return r.connect({host:serverHost, port:driverPort, authKey: "hunter-bad"});
                }).then(function() {
                    err = new Error('Unexpectedly connected with bad AuthKey');
                    done(err);
                }).catch(givesError("RqlDriverError", "Server dropped connection with message: \"ERROR: Incorrect authorization key.\"", done))
            });
        });
        
        // -- use tests
        
        this.timeout(8000) // "test use" can take some time
        
        it("test use", withConnection(function(done, c){
            r.db('test').tableCreate('t1').run(c, noError(function(){
                r.dbCreate('db2').run(c, noError(function(){
                    r.db('db2').tableCreate('t2').run(c, noError(function(){
                        c.use('db2');
                        r.table('t2').run(c, noError(function(){
                            c.use('test');
                            r.table('t2').run(c, givesError("RqlRuntimeError", "Table `test.t2` does not exist", done));
        }));}));}));}));}));
    });
});

File: test/rql_test/connections/connection.rb
##
# Tests the driver API for making connections and excercizes the networking code
###

$LOAD_PATH.unshift '../../drivers/ruby/lib'
$LOAD_PATH.unshift '../../build/drivers/ruby/rdb_protocol'
require 'rethinkdb'
extend RethinkDB::Shortcuts

require 'rubygems'
gem 'test-unit'
require 'test/unit'

$server_build_dir = ARGV[0]
$use_default_port = ARGV[1].to_i == 1

class TestNoConnection < Test::Unit::TestCase

  include RethinkDB::Shortcuts

  # No servers started yet so this should fail
  if $use_default_port
    def test_connect
      assert_raise(Errno::ECONNREFUSED){ r.connect }
    end
  end

  def test_connect_port
    assert_raise(Errno::ECONNREFUSED){ r.connect '0.0.0.0', 11221 }
    if $use_default_port
      assert_raise(Errno::ECONNREFUSED){ r.connect '0.0.0.0' }
    end
  end

  def test_empty_run
    # Test the error message when we pass nothing to run and
    # didn't call `repl`
    assert_raise(ArgumentError){ r.expr(1).run }
  end
end

class RethinkDBServer
  include RethinkDB::Shortcuts

  attr_reader :conn
  attr_reader :port

  @@rdb_count = 0

  def initialize(port = 0)
    if port == 0 then
      @port = rand(65535-1024)+1024
    else
      @port = port
    end
    run_dir = "run/server_#{$$}"
    Dir.mkdir run_dir if not File.exist? run_dir
    index = @@rdb_count += 1
    @io = IO.popen("#{$server_build_dir}/rethinkdb --http-port 0 --cluster-port 0 --driver-port #{@port} 2>&1 --directory #{run_dir}/rdb_#{index}")
    lines = []
    @io.each do |line|
      lines << line
      if line =~ /Server ready/
        @conn = r.connect 'localhost', @port
        return
      end
    end
    lines.each{ |line| puts "#{line}" }
    raise Exception, "RethinkDB process failed to start correctly"
  end

  def query q
    q.run(@conn)
  end

  def close
    @conn.close
    Process.kill "INT", @io.pid
    lines = @io.readlines
    @io.close
    if $? != 0
      lines.each{ |line| puts line }
      raise Exception, "RethinkDB process returned non-zero exit status"
    end
  end
end

if $use_default_port
  class TestConnectionDefaultPort < Test::Unit::TestCase
    include RethinkDB::Shortcuts

    def setup
      @server = RethinkDBServer.new 28015
    end

    def teardown
      @server.close
    end

    def test_connect
      conn = r.connect
      conn.reconnect
    end

    def test_connect_host
      conn = r.connect 'localhost'
      conn.reconnect
    end

    def test_connect_host_port
      conn = r.connect 'localhost', 28015
      conn.reconnect()
    end
  end
end

class TestWithConnection < Test::Unit::TestCase
  include RethinkDB::Shortcuts

  def setup
    @server = RethinkDBServer.new
    @conn = @server.conn
  end

  def teardown
    if @server != nil
      @server.close
    end
  end

  def query q
    q.run @conn
  end
end

class TestConnection < TestWithConnection

  def test_connect_close_reconnect
    query r.expr(1)
    @conn.close
    @conn.close
    @conn.reconnect
    query r.expr(1)
  end

  def test_connect_close_expr
    query r.expr(1)
    @conn.close
    assert_raise(RuntimeError){ query r.expr(1) }
  end

  def test_db
    query r.db('test').table_create('t1')
    query r.db_create('db2')
    query r.db('db2').table_create('t2')

    # Default db should be 'test' so this will work
    query r.table('t1')

    # Use a new database
    @conn.use('db2')
    r.table('t2')
    assert_raise(RethinkDB::RqlRuntimeError){ query r.table('t1') }

    @conn.use('test')
    r.table('t1')
    assert_raise(RethinkDB::RqlRuntimeError){ query r.table('t2') }

    @conn.close

    # Test setting the db in connect
    @conn = r.connect 'localhost', @server.port, 'db2'
    query r.table('t2')

    assert_raise(Exception){ query r.table('t1') }

    @conn.close

    # Test setting the db as a `run` option
    @conn = r.connect 'localhost', @server.port
    r.table('t2').run(@conn, 'db2')
  end

  def test_use_outdated
    query r.db('test').table_create('t1')

    # Use outdated is an option that can be passed to db.table or `run`
    # We're just testing here if the server actually accepts the option.

    r.table('t1', { :use_outdated => true }).run(@conn)
    r.table('t1').run({ :connection => @conn, :use_outdated => true })
  end

  def test_repl

    # Calling .repl() should set this connection as global state
    # to be used when `run` is not otherwise passed a connection.
    c = @conn.repl

    r.expr(1).run()

    c.repl() # is idempotent

    r.expr(1).run()

    @conn.close()

    assert_raise(RuntimeError){ r.expr(1).run }
  end
end

class TestShutdown < TestWithConnection
  def test_shutdown
    query r.expr(1)
    @server.close
    assert_raise(RuntimeError){ query r.expr(1) }
  end
end


# This doesn't really have anything to do with connections but it'll go
# in here for the time being.
class TestPrinting < Test::Unit::TestCase

    # Just test that RQL queries support __str__ using the pretty printer.
    # An exhaustive test of the pretty printer would be, well, exhausing.
    def runTest
        assert_equal("#{r.db('db1').table('tbl1').map(lambda{ |x| x })}",
                     "r.db('db1').table('tbl1').map(lambda var_1: var_1)")
    end
end
